### PR TITLE
Rework Actor and Systems

### LINKED ISSUE #1009: Improve Actor and Network actor by removing unnecessary event listeners
## Summary

There's no need to add PreUpdate and OnUpdate listeners if they are not being used.

### PR DESCRIPTION
**Pull request by joaoburatto**
## Summary

Reworks the whole Actor system to improve QOL and document long forgotten parts of that code.

- Renames the System concept to Subsystem, in a way to avoid conflicts with Unity's classes.
- Reworks the Actor class, adding documentation and an ActorLocator class to find actors via GameObject.GetInstanceID()
- Removes not-needed event registering on Actor
- Document most of the reworked classes.

### CODE REVIEW

**Comment by stilnat** [comment]
I won't take the responsibility to review this one. It's lacking a proper description, it's way too big, and most of the changes are not explained at all. A single sentence is clearly not enough to understand what's going on and why we even need that. As it is, I think merging it is a bad idea. 

What I expect before reviewing :
- Detailed account of the issues of the modified files, along with the proposed solution.
- Why do we need what was added.
- At the very least, every public class, structure, interface with a summary. Not talking about methods and fields.
- Remove the renaming so I can focus on what matters, make a PR for renaming.

We also already discussed about renaming stuff in a PR. This make the review process tedious as more than half the changes are just renaming, but the reviewer still has to be careful about it.

**Comment by cosmiccoincidence** [comment]
Is this PR meant to complete #1009 ??

**Comment by joaoburatto** [comment]
> Is this PR meant to complete #1009 ??

Good catch, I did not remembered I had created that one. Yes it does close it. Reasoning being that making all Actors listen to UpdateEvent is bad for performance if the OnUpdate callback is not used.

**Comment by joaoburatto** [comment]
> I won't take the responsibility to review this one. It's lacking a proper description, it's way too big, and most of the changes are not explained at all. A single sentence is clearly not enough to understand what's going on and why we even need that. As it is, I think merging it is a bad idea.
> 
> What I expect before reviewing :
> 
> * Detailed account of the issues of the modified files, along with the proposed solution.
> * Why do we need what was added.
> * At the very least, every public class, structure, interface with a summary. Not talking about methods and fields.
> * Remove the renaming so I can focus on what matters, make a PR for renaming.
> 
> We also already discussed about renaming stuff in a PR. This make the review process tedious as more than half the changes are just renaming, but the reviewer still has to be careful about it.

Updated the description. The renaming is a part of this PR.

File: Assets/Scripts/SS3D/Core/Behaviours/Actor.cs [hunk 1]
```diff
@@ -1,140 +1,140 @@
...
-        private bool _initialized;
+        public Transform Transform => GetTransform();
+        public GameObject GameObject => GetGameObject();
+        public RectTransform RectTransform => (RectTransform)Transform;
 
-        private readonly List<EventHandle> _eventHandles = new();
+        public bool ActiveSelf => GameObject.activeSelf;
+        public bool ActiveInHierarchy => GameObject.activeInHierarchy;
 
-        public Transform TransformCache
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Core/Behaviours/Actor.cs:22]: why is that removed ?

File: Assets/Scripts/SS3D/Core/Behaviours/Actor.cs [hunk 2]
```diff
@@ -1,140 +1,140 @@
...
 
 namespace SS3D.Core.Behaviours
 {
     /// <summary>
-    /// Used to optimize all GameObjects, avoid MonoBehaviours
+    /// Used to optimize all GameObjects, avoiding MonoBehaviours
     /// </summary>
     [Tooltip("Used to optimize all GameObjects, avoid MonoBehaviours")]
-    public class Actor : MonoBehaviour
+    public class Actor : ActorBase, IActor
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Core/Behaviours/Actor.cs:14]: please add a few lines to explain how it optimises Monobehaviour.

File: Assets/Scripts/SS3D/Core/Behaviours/NetworkActor.cs [hunk 3]
```diff
@@ -1,142 +1,135 @@
...
-using Coimbra.Services.PlayerLoopEvents;
-using FishNet.Object;
 using UnityEngine;
 
 // ReSharper disable InconsistentNaming
 namespace SS3D.Core.Behaviours
 {
     /// <summary>
-    /// Used to optimize all NetworkObjects, avoid MonoBehaviours
+    /// Used to optimize all GameObjects, avoid MonoBehaviours
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Core/Behaviours/NetworkActor.cs]: please add a few lines to explain how it optimises NetworkBehaviour.

File: Assets/Scripts/SS3D/Systems/Gamemodes/UI/GamemodeObjectivePanelView.cs [hunk 4]
```diff
@@ -69,6 +73,23 @@ private void HandleRoundStateUpdated(ref EventContext context, in RoundStateUpda
             }
         }
 
+        private void HandleUpdate(ref EventContext context, in UpdateEvent updateEvent)
+        {
+            ProcessInput();
+        }
+
+        private void ProcessInput()
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Systems/Gamemodes/UI/GamemodeObjectivePanelView.cs:82]: shouldn't that use the new input system ?

File: Assets/Scripts/SS3D/Systems/Roles/RoleSystem.cs [hunk 5]
```diff
@@ -17,11 +17,11 @@
 
 namespace SS3D.Systems.Roles
 {
-    public class RoleSystem : NetworkSystem
+    public class RoleSystem : NetworkSubsystem
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Systems/Roles/RoleSystem.cs]: rename to roleSubSystem for consistency.

**Review by stilnat — CHANGES_REQUESTED** [review]
A few questions, a few comments, nothing much to add.

File: Assets/Scripts/SS3D/Core/Behaviours/Actor.cs (hunk 1 above)
Comment (joaoburatto) [reply Assets/Scripts/SS3D/Core/Behaviours/Actor.cs:22]: Moved to `ActorBase.cs`

File: Assets/Scripts/SS3D/Systems/Gamemodes/UI/GamemodeObjectivePanelView.cs (hunk 4 above)
Comment (joaoburatto) [reply Assets/Scripts/SS3D/Systems/Gamemodes/UI/GamemodeObjectivePanelView.cs:82]: Unrelated to this PR. Only changed due to Actor changes.

**Comment by stilnat** [comment]
fix that stuff : 
image

**Comment by cosmiccoincidence** [comment]
We're the errors fixed? Is this ready for another review @joaoburatto ?

EDIT: Apparently not..

**Comment by cosmiccoincidence** [comment]
lgtm

**Review by stilnat — CHANGES_REQUESTED** [review]
Some new issues : Looks like the material for ghost construction stuff is out 
image

it broke something on the interaction system, I can't interact with some stuff anymore. Issues are not limited to the toolbox : 
image

**Comment by stilnat** [comment]
This PR modifies half the scripts of our repo, some changes are pretty significant, some were breaking systems, and although resolved, I can't give any guarantee that some of those other changes are not harmful. There's just too much to properly review it that I don't understand. I would expect changes to be solely present in Core.

My advice on this one is, close the PR, port only the relevant change on another one, please remove the renaming, it just affect too many files and it's painful to review when one has to be careful if it's really only renaming. Open a PR for renaming only after. I'm keeping "request changes" for now because I don't want it to be accidentally merged but if someone wants to review it, let me know if you approve it and I'll remove the request changes.

**Comment by stilnat** [comment]
Among the issues still present on this PR
- can catch items from the other side of the map
- materials in tilemap editor menu are absent (everything's purple)

**Comment by stilnat** [comment]
Why I believe this PR should not be merged : 

1. most changes are completely unrelated to the PR's topic, actor's improvement.
2. A few changes, completely unrelated to the PR's topic, were game breaking and a few more could be that were not spotted.
3. It's making very hard to identify what commit caused what bugs since it's so big, making future debugging painful.
4. John made important changes to systems he himself said know almost nothing about such as the inventory system and the UI around it.
5. It modifies code style in ways we did not agree upon collectively, check attachedcontainer.cs as an example, attributes are on two different lines in this PR, when they were on a single line previously. It's an arbitrary choice, not relying on our C# style guide, and it's very debatable if that's actually something we want. 
6. It's making it hard to focus on the actual point of the PR, improving actors. After all the time I spent on that, I still have no idea how it improves it really since I had to deal with all the unwanted changes.
7. It shuffle files and folders around with no explanations.

All in all, I feel like this PR is way, way below our standards, it's making everyone lose time, fixing bugs that should not exist in the first place, it lacks explanation on most of the changes, it's compromising our chances to relate a bug to a given commit, it's not well tested, and it's hard to understand. 

If John didn't try to modify half the repository, this would be probably merged after 3 commits and some discussions about the actual improvement. What frustrate me the most is, that this probably contains actual improvement.

Just to be very clear about this, I tried to stay as factual as possible, my only care is the well being of the game. Some of this might sounds harsh and it is, but as the lead dev, I have higher expectation for John than for other contributors. I would not let that PR pass for any other contributors, so I've no reason to do it for John.

That said, if another reviewer wants to merge it, I'll accept it too, but not before asking them if they read this beforehand, as I want them to fully understand the issues with merging this kind of PR.

**Pull request closed without merging**